src/main/kotlin/org/rust/ide/highlight/RsHighlightExitPointsHandlerFactory.kt

```diff
@@ -0,0 +1,105 @@
+            return RsHighlightExitPointsHandler(editor, file, target)
+        }
+        if (elementType == IDENTIFIER) {
+            val parent = target.parent
+            if (parent is RsMacroCall && (parent.referenceName == "panic" || parent.referenceName == "try")) {
+                return RsHighlightExitPointsHandler(editor, file, target)
+            }
+        }
+        if (testForExprBeforeBlock(target)) {
+            return RsHighlightExitPointsHandler(editor, file, target)
+        }
+        return null
+    }
+
+    private fun testForExprBeforeBlock(target: PsiElement): Boolean {
+        var context: PsiElement? = target
+        while (context != null) {
+            val parent = context.parent
+            if (context is RsExpr && parent is RsBlock) {
+                return true
+            }
+            context = parent
+        }
+        return false
+    }
+
+    private class RsHighlightExitPointsHandler(editor: Editor, file: PsiFile, var target: PsiElement) : HighlightUsagesHandlerBase<PsiElement>(editor, file) {
+        override fun getTargets() = listOf(target)
+
+        override fun selectTargets(targets: MutableList<PsiElement>?, selectionConsumer: Consumer<MutableList<PsiElement>>?) {
+            selectionConsumer?.consume(targets)
+        }
+
+        private fun getFunctionOrLambda(target: PsiElement): PsiElement? {
+            var context: PsiElement? = target
+            while (context != null) {
+                if (context is RsFunction || context is RsLambdaExpr) {
+                    return context
+                }
+                context = context.parent
+            }
+            return null
+        }
+
+        override fun computeUsages(targets: MutableList<PsiElement>?) {
+            val function: PsiElement? = getFunctionOrLambda(target)
+            function?.acceptChildren(object : RsVisitor() {
+                override fun visitElement(element: PsiElement?) {
+                    element?.acceptChildren(this)
+                }
+
+                override fun visitFunction(o: RsFunction) {}
+                override fun visitLambdaExpr(o: RsLambdaExpr) {}
+                override fun visitRetExpr(o: RsRetExpr) = addOccurrence(o)
+                override fun visitTryExpr(o: RsTryExpr) = addOccurrence(o)
+
+                override fun visitTryMacroArgument(o: RsTryMacroArgument) {
+                    val macroCall = o.parent as? RsMacroCall ?: return
+                    addOccurrence(macroCall)
+                }
+
+                override fun visitFormatMacroArgument(o: RsFormatMacroArgument) {
+                    val macroCall = o.parent as? RsMacroCall ?: return
+                    if (macroCall.referenceName == "panic") {
+                        addOccurrence(macroCall)
+                    }
+                }
+
+                override fun visitExpr(o: RsExpr) {
+                    if (o.parent is RsBlock) {
+                        addOccurrence(o)
```

Comment:
The problem I want to address stems from the fact that in real projects, expressions are not so short and simple as you use in tests and in examples. Try to open a couple of real projects and move the cursor through the code. You'll see that very often the exit point highlighting spans huge parts of functions. One-liners and "one-expression"ers are always 100% highlighted. Here's an example from cargo:

<img width="587" alt="screen shot 2017-07-24 at 18 14 27" src="https://user-images.githubusercontent.com/2101250/28530391-549d5fe0-709c-11e7-85c5-87775620cf15.png">

It's just too much noise for the functionality you need only occasionally.

Moreover, this highlighting now subdues another very useful feature &mdash; symbol occurrence highlighting. Let's assume I want to find all the occurrences of the `self` argument in a function, so I put the cursor to the nearest `self` (another example from cargo):

<img width="486" alt="screen shot 2017-07-24 at 18 16 41" src="https://user-images.githubusercontent.com/2101250/28530486-ad38ee44-709c-11e7-8b5c-7a3746b10b8b.png">

Here, a) most of the function is green because of too eager exit points highlighting, so I don't see `self`s, and b) the first `self` occurrence is not highlighted at all.

As the result, the occurrences highlighting feature effectively becomes non-working. It doesn't look like a reasonable trade-off for having exit points highlighting. That's why I think we need to be careful here and try to find a better solution.

The Groovy-solution I described above is devoid of these problems because it doesn't highlight exit points when the cursor is placed inside an expression. Only when it's placed on keywords (`return`, `throw`). In Rust, things are more complicated because, for example, `panic!()` is both an exit point highlighting trigger and a valid part of an expression. What should we do when someone places the cursor on `panic!()`? Highlight exit points? But maybe I meant to highlight all the `painc!()`s in the file and don't want to see `return`s, `?`s and other exit-stuff. Let alone diverging functions (`fn foo() -> !`) that are also exit-points, but not covered in the current implementation.

So, the feature is much more complicated than it might seem at the first glance. My opinion is that it's better to start with something simple and safe (the Groovy-style).